text/0000-project-descriptor-images.md

```diff
@@ -0,0 +1,159 @@
+# Meta
+[meta]: #meta
+- Name: Support specifying images in project descriptor
+- Start Date: 2021-10-26
+- Author(s): [@phil9909](https://github.com/phil9909), [@modulo11](https://github.com/modulo11), [@pbusko](https://github.com/pbusko), [@jkutner](https://github.com/jkutner)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Add a new array of tables `[[images]]` to the `project.toml` file.
+
+# Definitions
+[definitions]: #definitions
+
+* **project descriptor**: The `project.toml` file that describes a project.
+* **image**: the output of a running a buildpack(s) against a project
+* **image name**: fully qualified name of an OCI image in the form `<registry>/<name>:<tag>`.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, there is no support for specifying an output image in the `project.toml` although flags like `--tag` are already present.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Buildpack user will be able to specify the resulting images in the `project.toml`:
+
+# How it Works
+[how-it-works]: #how-it-works
+
+- Accepted project descriptor related RFCs
+  - https://github.com/buildpacks/rfcs/blob/main/text/0019-project-descriptor.md
+  - https://github.com/buildpacks/rfcs/blob/main/text/0054-project-descriptor-schema.md
+  - https://github.com/buildpacks/rfcs/blob/main/text/0080-builder-key-in-project-descriptor.md (could serve as a role-model)
+  - https://github.com/buildpacks/rfcs/blob/main/text/0084-project-descriptor-domains.md
+  - Closed PR which includes some similar changes: https://github.com/buildpacks/rfcs/blob/project-descriptor-simple/text/0000-project-descriptor.md
+
+Example:
+
+```toml
+[[images]]
+name = "spring-petclinic"
+registry = "gcr.io"
+tags = ["latest", "v1"]
+
+[[images]]
+name = "spring-petclinic"
+tags = ["latest", "v1"]
+
+[[images]]
+name = "spring-petclinic"
+registry = ["private.registry.corp:8443"]
+```
+
+This example will produce the following images:
+
+- `gcr.io/spring-petclinic:latest`
+- `gcr.io/spring-petclinic:v1`
+- `docker.io/spring-petclinic:latest`
+- `docker.io/spring-petclinic:v1`
+- `private.registry.corp:8443/spring-petclinic:latest`
+
+# Alternatives
+[alternatives]: #alternatives
+
+- have `registries` as an array:
+
+  ```toml
+  [images]
+  registries = ["gcr.io", "docker.io"]
+  name = "spring-petclinic"
+  tags = ["latest", "v1"]
+  ```
+
+- have `tag` as a string instead of `tags` array:
+
+  ```toml
+  [[images]]
+  registry = "gcr.io"
+  name = "spring-petclinic"
+  tag = "latest"
+
+  [[images]]
+  registry = "docker.io"
+  name = "spring-petclinic"
+  tag = "v1"
+
+  # ...
+  ```
+
+- Simple array of strings containing all information i.e. `registry`, `name`, `tag`
```

Comment:
I'm more of a fan of this alternative. It's versatile enough to allow for any possible desires that I can think of while maintaining simplicity. The only cost I see is the potential repetitiveness of the host/registry. Are there other drawbacks/benefits to consider between this and the main proposed solution?

Reply:
One of the drawbacks is that we will loose the possibility to apply defaults to certain keys as described in the lines 141-155, e.g. to fall back to the default `docker.io` registry if none is specified, or to apply the `latest` tag.
Also apart from the potential repetitiveness of the host/registry combinations, the main proposed solution is probably easier to consume by the platform, which would simplify for example a validation of inputs.

Reply:
As long as the registry, name, and tag are structured in a url scheme (and not arbitrary) then wouldn't it still be possible for a value like `"me/myimage"` to be converted to `"docker.io/me/myimage:latest"`?

Reply:
You're right, it's possible to do it also with the array of canonical names, but on the other hand image tags are frequently being accessed/changed by that CI systems which produces the actual build. Having this data structured makes it a lot easier to consume/modify programatically.

Reply:
The problem I foresee with structured format is that I don't think there's a convention in the OCI realm on what `registry` vs `name` vs `tag` is. We're getting there but it may still be confusing to newcomers.

For example, on `grc.io` with image `gcr.io/cncf-buildpacks-ci/demo/kpdemo:v1`: 

- What is the `registry`, `gcr.io` (hostname) or `gcr.io/cncf-buildpacks-ci/demo`?
- What is the `name`, `cncf-buildpacks-ci/demo/kpdemo` or `kpdemo`? 
- What is the tag, `gcr.io/cncf-buildpacks-ci/demo/kpdemo:v1` or `v1`?
   - (the answer sounds obvious, `tag` but many systems, even `pack` use `tag` to mean the full reference)

Do your answers change on `docker.io` with image `index.docker.io/library/bash:latest`?

Reply:
According to the Docker Container Trust [documentation](https://docs.docker.com/engine/security/trust/#image-tags-and-dct), the Global Unique Name is composed as following:
`[REGISTRY_HOST[:REGISTRY_PORT]/]REPOSITORY[:TAG]`

This structure is also used by Notary, Cosign and google/go-containerregistry. In such case the answers are:

* registry is `index.docker.io` / `gcr.io`
* repository (name) is `library/bash` / `cncf-buildpacks-ci/demo/kpdemo`
* tag is `latest` / `v1`

The main confusion could be coming from the usage of `--tag` argument (`pack` refers to the GUN, instead of the last part after the `:`)

Reply:
Thank you for those learnings. I can concede to the use of a GUN. 😉 

I suppose another reason for the reservation has to do with considering other images that go into a Buildpacks build project. I do wonder if we should look at this holistically. For example, `run` image isn't hard to envision being part of the project descriptor. How would that look like side by side? Why isn't the [builder](https://github.com/buildpacks/rfcs/pull/138) property decomposed?

[Here's](https://github.com/buildpacks/rfcs/blob/e9f36ba8718926edb62652ccc1767ebfcd26eef1/text/0000-buildpacks-config.md?plain=1#L122-L140) a _very_ different format on this proposed configuration. I wonder how that would look if certain image references had a "decomposed" structure.

Reply:
@jkutner, I conceded to the idea of GUN but still not convinced that we should be doing the component level breakdown in the configuration. I'd like to have more discussion around this since it would have implications _anywhere_ we reference an image.

Reply:
To elaborate a little, I would like to request that whatever decision be made here that it apply to all image references. This would include making the same decomposition to `builder` and any future references to images. 

In my opinion, the decomposition doesn't make sense everywhere such as `builder`, `run-image` (very likely future parameter). The cost/benefit just doesn't add up for me.